File: Data Structures & Algorithms/eating-bananas/submission-4.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        l = 1
        r = max(piles)
        result = r

        while l <= r:
            k = (r + l) // 2

            total = 0
            for pile in piles:
                total += math.ceil(pile/k)

            if total > h:
                l = k + 1
            else:
                result = k
                r = k - 1

        return result
        

File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-4.js
class Solution {
    /**
     * @param {string} s
     * @param {number} k
     * @return {number}
     */
    characterReplacement(s, k) {
        const map = {};
        let l = 0;
        let result = 0;
        let currMax = 0;

        for (let r = 0, n = s.length; r < n; r++) {
            map[s[r]] = (map[s[r]] || 0) + 1;
            currMax = Math.max(currMax, map[s[r]])

            while (r - l + 1  - currMax > k) {
                map[s[l]] -= 1;
                l++;
            }
            result = Math.max(r - l + 1, result)
        }
        return result
    }
}
